Add text & color label filters

phweyland opened the pull request:
Enhancement:
Adds a text filter which acts on all image text data: metadata, tags, filename and path.

Bug fix:
On every click / selection, the full collection query was executed two or three times. Now it reuses the memory.collected_images table instead.

Notes
I've found some difficulty in setting up an acceptable layout. Some strange behaviors I haven't explained:
- in darktable.c `min-width` looks like a minimum width but acts as a constant width.
- between the filter area and the 5 main buttons on the right, there is a not compressible zone, which could be used in case of small screen width.

Beside this, the usage of gtk_box_set_homogeneous() is not very convenient for fields of different width EDIT I've removed it but I'm not sure that's right.

AlicVB commented:
I've tested extensively, but I really like the idea of a filter by "search among all metadatas". I already can see many use of that...
a little point, I would have put the search box before the sort, so the 2 "filters" stay together.

My only fear is that we have other very good candidates for the "quick filters", like colorlabels for example (see #6675). I fear that all this will take too much space at some point. not to speak about the UI cluttering...
(but that's another story... maybe one time someone will have a better design for all that collect-filter-sort part...)

phweyland commented:
> a little point, I would have put the search box before the sort, so the 2 "filters" stay together.

Right, I make a try.

quovadit commented:
it would be great if "quick filters" in lighttable were as customizable as the "quick access panel" in darkroom.

So to avoid UI cluttering, the user could define often-used search-criteria (rating, color-label, full-text, ...) to be visible in the top-bar.

dterrahe commented:
Could you give a brief rationale for implementing this as a separate search layer instead of adding it as a pseudo field ("any text") to the normal collections lib filter options list? That would have given the possibility to _exclude_ on any given text or _combine_ several text searches which this doesn't have. Or include it in a saved preset.

I can see the benefit of a quick&dirty search, available in darkroom, but as @AlicVB points out, that probably applies to other fields as well.

phweyland commented:
> Could you give a brief rationale for implementing this as a separate search layer instead of adding it as a pseudo field ("any text") to the normal collections lib filter options list?

That was my first intention. Adding a PROP entry should be easy. But then how to show the result ? You can have some fit on path, metadata and tags at the same times. Apart showing the fitting images in the view center, how to organize the collect module view ? My neurons make a knot here.
My goal was also more volatile than really creating a new collection. Just to find something I don't remember  where to search. I select the upper folder level I'm sure about (may be the root) in collect module and type in the desired text (not at all the presets thing).

dterrahe commented:
I hope it is obvious that the reason I asked is that it is foreseeable that when this useful feature gets added, people will immediately start asking for it to be added to the collect filters as well, to have the benefits I indicated. So if it could possibly be implemented in such a way that functionality doesn't need to be duplicated at a later stage, that would be good.

Possibly by adding this to collect and then taking @quovadit's suggestion of having quick filters for any of the options in collect. Maybe we don't need the full power to be able to add all of them at the same time (then we would also need to define combine and exclude options); it might be sufficient to have a single type dropdown+value (like the first line in collect) that would then implicitly _narrow_ the current set. This would basically provide you with the functionality you have here (if "any text" is the default filter).

phweyland commented:
> people will immediately start asking for it to be added to the collect filters as well,

This may or not be true. This text filter is particularly inefficient in term of performance. And the idea itself is more a curative than an organization method. 


> adding this to collect

Thinking more about it, it comes to me that such a "filter option item" in collect module could be allowed only if not at the first level (some boring coding here). For example, the first level is folder, and one adds a second level with such a text filter. In that case, the collect module shows the folders containing images which fit with the filter.
But, from the UI's standpoint the quick filter as proposed here seems easier to use.

TurboGit commented:
I'm wondering if for such a generic search it would not be better to not use % as wildcard but have the typed text be used a sub-string (that is always use %typed-text%) and also maybe start the search at each typed character.

TurboGit commented:
I also agree with @dterrahe that time will come where we will want this part of the collection module.

quovadit commented:
> how to organize the collect module view ? My neurons make a knot here.

you could show the results as hierarchy, analog to a search in tags, with the field as top-element:
- filename:
  - xmas-party_01.jpg (1)
  - xmas-party_02.jpg (1)
- tag:
  - event
    - xmas (42)
- title:
  - Xmas Party 2021 (37)



> such a "filter option item" in collect module could be allowed only if not at the first level

I don't understand why this would be necessary, and I don't think it is preferable as a user.
Maybe I first want to search, and then narrow down by camera, date, ...

TurboGit commented:
> you could show the results as hierarchy, analog to a search in tags, with the field as top-element:

But I understand @phweyland.... This seems something that will be tricky and complex to implement! So we need really to come with a good idea of what is behind us and drop this idea if it turns too complex.

phweyland commented:
> it would not be better to not use % as wildcard but have the typed text be used a sub-string (that is always use `%<typed-text>%`)

That's the easy way, true, but it can (will) produce more results than expected, and the user cannot fix that. I prefer to keep the control, but that can be a preference setting.

>  start the search at each typed character.

As said above, the use-case I think about is to find something I've lost (I don't remember which attributes have been set). Then, most of the time it applies to the whole library and the query is slow.

> time will come where we will want this part of the collection module.

Sure someone will want that. But what is the use case ?
The collect module is powerful but not easy to use. Mixing easy navigation and powerful filtering is not an easy task. Cluttering it a bit more may not be the way to go. 
Filtering in collect module makes sense when it can be useful to reuse it (through saved presets). For quick searches which don't deserve to be saved, the quick search toolbar is better.
Of course if somebody proposes an attractive solution with collect module it is always possible to move to it.

> you could show the results as hierarchy, analog to a search in tags, with the field as top-element:

Showing different structures in the same tree requires specific code in collect.c and collection.c as well.
I think it's worth to understand why we should do that (which use-case ?).

phweyland commented:
colorlabels filter not yet finished. Trying to solve #6675.

phweyland commented:
Now functional for me but not tested on large database.

quovadit commented:
works great, just one finding:

If you exclude more than one color (e.g. ctrl-click red + yellow) and switch to union-mode, all images except red+yellow are shown, but the rating-filter is ignored, so I see rejected, 1-star, ... although rating filter is set to >= 4 stars.

phweyland commented:
I think I understand why... I'll try to fix this asap. Thanks for testing!

phweyland commented:
> If you exclude more than one color (e.g. ctrl-click red + yellow) and switch to union-mode, all images except red+yellow are shown, but the rating-filter is ignored, so I see rejected, 1-star, ... although rating filter is set to >= 4 stars.

This seems to work better now.

quovadit commented:
yes, works perfect, thanks a lot.

I would just adapt the tooltip for the union/intersect toggle:

intersection (∩): only images having ALL selected colors
union (∪): all images with at least ONE selected color

phweyland commented:
rebased for conflicts with #10768

TurboGit commented:
And again some conflicts, hope those will be easier to resolve. TIA.

Nilvus commented:
I missed that one. I'm quite sure it's not the good way to go. It's really great to have such text filter but this clutter the top bar. And I really think it's not the good place, nor the best way. We need to consider the work from @AlicVB about a new filter collections module and the idea, previously started by @aurelienpierre to think about a better way to use top bar.

So we need to think about how all things come between them. I'm afraid that with actual proposals, the UI that started to be better, will come back in the older way.

Apart of that, for just that PR, I see some downsides:

- having a text filter alone like that in top bar, is just bad UX/UI design, and absolutely not intuitive (we don't know why there's a search box there)
- having color labels that way do not differentiate them regarding of same on the bottom. Which one to do what?
- and if we have too many filters on the top bar and a new filter collections module, this will be harder to understand how they work together.

Those downsides are just from a new user point of view.

I think we need to spend time to think about a better way with all those proposals about improving DAM parts (and it's needed).

Form my point of view, I think improving at first filters in collections modules by helping @AlicVB improving its proposals is the place to start. Then after having a good new filters collections module, we could think about how to improve the top bar, and what to do about it. Of course, text filter here should be added in @AlicVB filters module.

Last, on what I talk about @aurelienpierre proposal, you could remember its work started here: https://github.com/darktable-org/darktable/pull/9015

And on that PR, that could mean moving filters on another place, depending on what we do/add later. And of course, depending of what sort of course become.

AlicVB commented:
As I've already said "I don't remember where" I think that one solution could be to code everything in the filtering module (a one like in #11066) and then let the user choose some filters to mirror in the toolbar. A little bit like what we do we QAP. My work on new widget design is done with that possibility in mind.
@Nilvus : sadly removing completely the filtering toolbar in lighttable is not really possible as we would break the workflow for user that hide side panels to maximize the space for thumbs. But what can be possible is to offer the possibility the choose which filters to show there (or to show nothing)
For darktable, that's another story, and I am in favor to remove it completely, as anyway it's only related to filmstrip and very far from it, so confusing...

quovadit commented:
I love the new text filter, and I don't think it clutters the top bar at all.
But I think, that the whole filter bar should be limited to the lighttable, because darkroom has a different focus.

For me it's perfect UX to have a search bar above a list of images, it's very clear that you can filter them by searching for different criteria. And there is a tooltip explaining exactly what it does.
But to make it a simple-to-use filter, I would always do a wildcard search without having to add '%' before and after the search term.

I also don't see a problem having the color labels on top and at the bottom. The rating stars have always been at both locations, and it's clear that you can filter on top, and assign at the bottom.
And again: there are tooltips.

I agree, that the new filter bar has to be coordinated with the new collections-module.
The perfect solution would be the already-mentioned QAP-style.

phweyland commented:
> But to make it a simple-to-use filter, I would always do a wildcard search without having to add '%' before and after the search term.

Pascal has already suggested that. And I agree that would often easier. But if you want the images having the text start with "xxx", they get mixed with all images matching the text.
I can do the following. Enter to search with implicit starting and ending %. Ctrl+Enter for explicit %.

quovadit commented:
> But if you want the images having the text start with "xxx", they get mixed with all images matching the text.

you could also use quotes for exact searches without auto-wildcards, so "xxx%" for everything starting with xxx

phweyland commented:
- Type text + Enter => automatic start and end %
- Type text + Home + Enter (cursor at pos 0) => no automatic start and end %. The user put the % where he needs.

Ctrl+Enter for the second case is also possible but the code is just less simple.

TurboGit commented:
To balance in favor of this one could see this as the search for a module name. Some editors also have those kind of search facility. And here we are more talking about "quick filtering" and not something that will use to create collection. All in all I'm not opposed to that and I have no strong opinion for or against it. That being said I would love to have the filtering works as done for modules, that is as soon as a character is entered. That's also how works the omniscient search in a Web page using Firefox or Chrome for example.

Nilvus commented:
> @Nilvus : sadly removing completely the filtering toolbar in lighttable is not really possible as we would break the workflow for user that hide side panels to maximize the space for thumbs. But what can be possible is to offer the possibility the choose which filters to show there (or to show nothing)
> For darktable, that's another story, and I am in favor to remove it completely, as anyway it's only related to filmstrip and very far from it, so confusing...

I'm absolutely not for removing filter top bar. I just remember here that was something discussed some times ago. And so, if we add things on that top bar, that means we should all agree that this top bar will remain, at least in lighttable. And yes, I find that such quick filter make sense in lighttable, a lot less in darkroom. 
We just need to be sure that this won't change soon, that could confuse more users.

My comment here was hard, I'm sorry for that. That's not only this PR but the fact that we've had too much proposals about those filters and top bar and sometimes they are not compatible. That show one main issue: we lack some global design thinking. I really think that such filters need more discussion to find a better way.

I've thought more about that PR since my answer and after reading your reactions and maybe adding that is not bad but I'm sure they're something remaining that is not good UX/UI here. Maybe not the fact that they are here but how they are displayed. Maybe some UI improvements (could be details) would make those better.

@quovadit: it's good if it's ok for you but, I already remember it to you some days ago in another PR, improving the UI is not only about you (or even me) but what could be the best compromise for most of the users. And we need too to think about new users.

> But I think, that the whole filter bar should be limited to the lighttable, because darkroom has a different focus.

Of course, I completely agree here!
 
> For me it's perfect UX to have a search bar above a list of images, it's very clear that you can filter them by searching for different criteria. And there is a tooltip explaining exactly what it does. But to make it a simple-to-use filter, I would always do a wildcard search without having to add '%' before and after the search term.

I disagree the way it is displayed now. It's not very clear, maybe as simple find text would be better.

> I also don't see a problem having the color labels on top and at the bottom. The rating stars have always been at both locations, and it's clear that you can filter on top, and assign at the bottom. And again: there are tooltips.

Yes, they have always been at the botom to set them. But that argument works for old users, not the new users that will start using darktable with next release (and so this filters).

And tooltips could be deactivated. Tooltips are useful but it's better if a function could be understood (at least, it's main use) without moving the mouse hover it (so without seeing tooltip). That's better UX design. Even if I could make errors, I think I prove some UX design skills.

> I agree, that the new filter bar has to be coordinated with the new collections-module. The perfect solution would be the already-mentioned QAP-style.

Completely agree here!

> The collect module is powerful but not easy to use. Mixing easy navigation and powerful filtering is not an easy task. Cluttering it a bit more may not be the way to go.

Clearly and that's the goal of @AlicVB to make that easier. Difficult work. One reason that not make it easy for some people is the use of booleans for filters conditions. So, I'm afraid that using boolean, even if they are 2 and quite simple one (and even if there's a tooltip), is not a good idea. I of course talk about those ones:

_intersection (∩): only images having ALL selected colors
union (∪): all images with at least ONE selected color_

Talking about @AlicVB and some asking here about adding text filter, It could probably be added later when @AlicVB will be over. We will certainly have a better view on how this could be implemented (or not) depending on the final result).

As a reminder about such issue, read this old issue:  #9664

I here don't talk about me, such boolean is not an issue for me but I understand this could be for users with no maths or scientific background.

To resume, I'm completely with you here for:
- of course, keep top bar
- having customizable top bar (especially to let users have filters they want) but with limited options (you talk about that here that not all filters could be suitable for quick filters)
- such quick filters should be limited to lighttable view (maybe map one too...)

But, there something that is not completely good UI/UX design here. Even if I can't, by now, be sure for what and so how to improve that.

I remains afraid (I could be wrong about those fears) that some confusion for new users will remains at start between color labels on top and those on bottom. And search is actually not so intuitive on what it does (before reading tooltip). Maybe this last thing just need some little UI improvements.

Last: as I said, my reaction is mainly because UI is mainly made as little blocks separately and I think we need to improve that, especially for big work, big part that are not intuitive on darktable, like actual filters (I don't talk about quick filters here but about how darktable let the user filter images here, so mainly collections modules, but quick filters are related to that).

phweyland commented:
>  But, there something that is not completely good UI/UX design here. Even if I can't, by now, be sure for what and so how to improve that.

Reading this as a question, a clearer visual separation between blocks would certainly help the user.
Some good css set of rules (when and how to use class selectors, instead of adhoc id selectors) would help the devs to provide more consistent UI faster.

> I would love to have the filtering works as done for modules, that is as soon as a character is entered.

That's the dream indeed. 
Based on my tests on 50k images 2k tags db, the initial search takes something like 2 seconds. Then I haven't even try the parallel search. I'll do it and come back to you with some results.

TurboGit commented:
> That's the dream indeed.

Maybe start the search when 2 or 3 characters are entered? Or maybe it doesn't make any difference?

Also, I suppose this quick filtering is better intended to filter a collection not the whole database. But that's a matter of usage.

TurboGit commented:
> Then I haven't even try the parallel search. I'll do it and come back to you with some results.

Thanks!

Nilvus commented:
@phweyland: I've worked about testing your PR but by improving UI/UX and I've to admit that your PR could be great with some improvement. I propose this mockup:

![image](https://user-images.githubusercontent.com/45535283/153713256-cbee53ea-2839-4e62-b180-eef301e26711.png)

First, I suggest to add ":" and a little more space after view text to have so:

view: {little more space} then filters.

It's a minor thing but I think this could get "view" text more as a "title" of that PR and so improving the idea that all after is filter to display/view images. And so, reduce issues I pointed. I'm even wondering if this text could not be set bold but I'm not sure.

About search box, I think this should be set same way as search modules box in right panel on darkroom (by just adjusting background color if needed to suit more as background top bar. This will improve contrast. But also have same behavior for reset button (the cross), having it on search box but only display it if text is entered as search box in darkroom. Last adding text like proposed above. Of course, not so brighter as in this mockup but exactly as search box in modules darkroom part.

Is there a reason you keep grey icon as before and not using new grey icon I propose? I think this one could be used there too.

Last, all is badly align vertically, even my sortby icon, so not your PR. If all this could be better aligned as I do in above mockup.

Now remains boolean added and issue I pointed. I fear that some users would not like it. And I'm wondering if boolean is really useful here. It's possible to do same things but just toggle color labels. As we have only 6 ones, it remains easy to do the same without boolean (of course with just few more click but I think boolean use will not be used so often, so no real time lost and space saved without them).

Nilvus commented:
Oh, to better align vertically, it's better to remove header bar and just display top bar on top. It's text and combobox that are not correctly align vertically. Color labels, search box and sortby icon are ok.

See:

![image](https://user-images.githubusercontent.com/45535283/153713667-502a27fc-5f7e-4635-b851-ff751761c159.png)

AlicVB commented:
> when @AlicVB will be over.

what ??? all this just because I've asked you to code a little bit of C ??? :-D (at least I'm not the only one to make stupid typing mistake)

More seriously, imho : 
- top bar "quick filter" shouldn't provide boolean choice. Always "and"
- for doing more clever stuff, we'll have the "filtering" module (which will *not* be the collect module, see #11066)
- implementing a "QAP" style solution, is really needed here : as said earlier, useful filter heavily depend on user workflow. We just can propose some default configuration (star + colorlabels, maybe) and let the user define it to his liking. I've even a name for that :) "QFB" (quick filter bar)
- maybe we need some sort of separator between filters, or better to put each filters on a darker background like what is done in #11066

Nilvus commented:
> > when @AlicVB will be over.
> 
> what ??? all this just because I've asked you to code a little bit of C ??? :-D (at least I'm not the only one to make stupid typing mistake)

My english is limited. I only understand, after reading this 3 times, my stupid typing mistake... My bad. Of course, when you work will be finished, but you understand ;-)

For you proposal, I agree. But, at first step, this PR could be merged with little tweaks I propose (improving visually it) and then starting working on those proposal to improve that.

phweyland commented:
> > I would love to have the filtering works as done for modules, that is as soon as a character is entered.
> 
> That's the dream indeed.

That's not exactly the dream, but hitting Enter is not necessary any more. 
The entered text is dimmed while the query is executed (kind of warning that the entry is frozen). Not sure we can get better with a single thread.
The query doesn't start with less than 4 characters.
Hitting Enter can force the query in that case.
Home + Enter can still be used to force "without" wildcards.

TurboGit commented:
> The query doesn't start with less than 4 characters.
> Hitting Enter forces the query in the previous case.
> Home + Enter can still be used to force "without" wildcards.

I suppose this is a good compromise for all users. Thanks for working on this, I really feel it is a good improvement. Will test soon.

phweyland commented:
The same using the modulegroups code.
Code is simpler. The only drawback is the magnifying glass space is not reused.
![image](https://user-images.githubusercontent.com/23012047/153771569-1c1577c6-41fc-4322-96a6-b876c0aef7ee.png)
![image](https://user-images.githubusercontent.com/23012047/153771589-4f33fb8e-bbb7-4f3c-ae3e-39e96fce3417.png)

With the elegant grey theme, the background tooltip is not visible.
Maybe a darker text entry darker instead of a lighter one would be better.

phweyland commented:
![image](https://user-images.githubusercontent.com/23012047/153777159-4b3a5a61-a04d-4f16-84d9-58dc20724433.png)
![image](https://user-images.githubusercontent.com/23012047/153777173-e40ab9de-9d5e-4d5c-a8ad-e133a6853776.png)
![image](https://user-images.githubusercontent.com/23012047/153777188-5a3f30a5-b0a2-4519-9214-1a22fbd3b628.png)

TurboGit commented:
> ![image](https://user-images.githubusercontent.com/23012047/153777159-4b3a5a61-a04d-4f16-84d9-58dc20724433.png) ![image](https://user-images.githubusercontent.com/23012047/153777173-e40ab9de-9d5e-4d5c-a8ad-e133a6853776.png) ![image](https://user-images.githubusercontent.com/23012047/153777188-5a3f30a5-b0a2-4519-9214-1a22fbd3b628.png)

I don't like the box around all part. I preferred the UI in the comment just before.

phweyland commented:
Back here:
![image](https://user-images.githubusercontent.com/23012047/153781873-2f268e85-3efc-49b6-ba51-ad375c7587a5.png)
![image](https://user-images.githubusercontent.com/23012047/153781920-259070db-bc17-4cdf-a06b-cbb7a1e10b80.png)

In any case those are small adjustments. Can be tweaked at any time.

phweyland commented:
> * top bar "quick filter" shouldn't provide boolean choice. Always "and"
> 
> * for doing more clever stuff, we'll have the "filtering" module (which will _not_ be the collect module, see

I think that doesn't hurt to have the choice. Can we talk again about this when "filtering" module gets merged. We should have a clearer view of the context.
BTW I'm a bit lost here. I thought you had suggested to share the code between different filters. This is the opposite, right ?

> * implementing a "QAP" style solution,

By this you mean configurable filter bar, is that correct ?

quovadit commented:
> * "quick filter" shouldn't provide boolean choice. Always "and"

I think it's great to have the choice, and the the default is already 'and', so I don't see any disadvantages.

phweyland commented:
> * better to put each filters on a darker background

I think that could be good. If I can find the way I'll make a try.


> * We just can propose some default configuration (star + colorlabels, maybe) and let the user define it to his liking.

What would be place for this ? A new chapter in preferences / lighttable ?
Or through a button like hamburger preferences ?

AlicVB commented:
I don't like the "borders version" either. What I had in mind was more something like : ![Capture d’écran_2022-02-14_13-25-12](https://user-images.githubusercontent.com/1818223/153864466-e0918914-adf8-442b-aec3-3538a59d9091.png)
(with some refinement on heights, margin/paddings... The sort button should be included inside the "sort box" too...
This would allow to better differentiate the different filters imho...

Now that I have tested, I see that I've misunderstood the "and/or" part. Sorry for that : I thought that you were speaking about the interaction between the different rules, not inside the colorlabels one.
So yes, no pb for the one inside colorlabels, even if I find quite strange the empty space... maybe a very dimmed "and" icon ??? (not sure)

For the relation between the new filtering module and the toolbar, here is how I see that : 
- In the filtering module, allow users to "pinned into toolbar" filters.
- In filtering module, "pinned" filters will be "visually" slightly different that "normal" ones.
- In filtering module, "pinned" filters will have some controls disabled (main operator, property changing menu, ...) and needs to be harder to remove to avoid "accidents"
- If a filter is "pinned, a "mirrored" version of the filter is shown in the header bar. Eventually the widget used there can be "simplified" a little bit (by removing the "entry line"  if we speak about the ratio widget of #11066 for example...)
- that way everything occurs inside the filtering module (no code duplication) and the header toolbar becomes only a container for "mirrored" filtering widgets.
- this would allow users to set the filters they want/need for their workflow. This would also allow to remove completely the header toolbar if they want (and work only with the filtering module)
- of course we need to come with a minimal setup for first usage (I would say just rating, maybe colorlabels)

TurboGit commented:
> Or through a button like hamburger preferences ?

No speaking for @AlicVB but I suppose what he had in mind is a way to create presets like in all modules? The presets would store the current filter status.

phweyland commented:
> maybe a very dimmed "and" icon ??

agreed.

> if you need help, I can share the little diff and css...

It's just to make a sensible choice. Which @base_color have you used in your example ?

> For the relation between the new filtering module and the toolbar, here is how I see that : []

OK, so it's too early to anticipate on that. We need the filtering module up and running to align the quick filter bar, right ?

AlicVB commented:
> Which @base_color have you used in your example ?

@border_color



> OK, so it's too early to anticipate on that. We need the filtering module up and running to align the quick filter bar, right ?

yes, but the work here is not lost either, as at some point I will "just" move it to filtering module... The ideal for me would be to have this merged first, so I can shamefully steal and mess with your code without removing your original authorship :)

phweyland commented:
![image](https://user-images.githubusercontent.com/23012047/154084492-3a8b5c20-d8a0-4f51-a836-098d4940900f.png)
![image](https://user-images.githubusercontent.com/23012047/154084981-e895a024-5f13-429b-b60b-f608e72e878c.png)

Nilvus commented:
Just few more thoughts:

1) why "search text" and not "search photos" written in search box?
2) is a "double-clic" possible to add in colorlabels part (in any part) to reset them by default (as we already have in many parts of darktable) or too hard to implement in such part?
3) in darktable.css file, you put a ".dt-dimmed" id but it seems to correspond to nothing now (maybe was useful on one previous proposal). I didn't find it on any part with Gtk Inspector tool and changing opacity or set other parts like color, background-color had no visible effect. Probably no more needed
4) sortby icon is quite small now. I think and is not so small in actual master. Maybe a little bigger would be better. But this also could be tweaked after merging that PR and in some fine-tuning I could propose after. As you want.

phweyland commented:
> I see a big issue on that PR in other languages

Yes. I look at it.

> why "search text" and not "search photos" written in search box?

This filter searches for text as color filter search for colors, stars filter for stars...

> in darktable.css file, you put a ".dt-dimmed" id but it seems to correspond to nothing now

It is not an id selector but a class selector. As such it can be associated to any other class on any object (what an id selector cannot do).
In that specific case it dims the entered text while the entry is frozen by the query execution.
BTW, darktable.css is overloaded by selector ids. They are useful, sure, but should reserved to specific cases. Seems we have too many specific cases ! Class selectors could help implementing general rules... 
 
> sortby icon is quite small now. I think and is not so small in actual master.

I wouldn't say "quite" but it is smaller, true.
It can be bigger but in that case it forces the sorting box (and the other filter boxes) to be bigger as well.

phweyland commented:
> I see a big issue on that PR in other languages

fixed:
![image](https://user-images.githubusercontent.com/23012047/154140739-ec9b2c54-78bd-4f43-94cb-6ee58e006004.png)

Nilvus commented:
> This filter searches for text as color filter search for colors, stars filter for stars...

I understand that but a search box is for searching text, but on what/for what. I think "search on photos" would be better.

> It is not an id selector but a class selector. As such it can be associated to any other class on any object (what an id selector cannot do). In that specific case it dims the entered text while the entry is frozen by the query execution. BTW, darktable.css is overloaded by selector ids. They are useful, sure, but should reserved to specific cases. Seems we have too many specific cases ! Class selectors could help implementing general rules...

Of course, I mean class, not id. I know the difference, just a too quick writing.
Ok, I understand the use, thanks for your explanation.
About the fact that darktable.css is overload by selector id. I will try this year to see how to change/improve that (big work). There's already quite a lot of general rules but I agree that this need, again, improvements. I do some one year ago, probably time to find now to make another round!

> I wouldn't say "quite" but it is smaller, true. It can be bigger but in that case it forces the sorting box (and the other filter boxes) to be bigger as well. EDIT there may be a crop feature in css ... could remove the extra pixels around the icon.

Let it as it is on your PR. I will see after this PR is merged how to better balanced that (and other parts, with CSS (and paint.c if needed).

Nilvus commented:
If it's ok for all, I'm ok for merging this. I could work on improving some little parts on the UI (and start updating some id to class, to make general rules (starting by search box).

quovadit commented:
> I understand that a search box is for searching text, but on what/for what. I think "search on photos" would be better.

I totally agree, that 'search photos' would be better. It's clear that a textbox is for entering text, but it's not obvious what you are searching for:

> we don't know why there's a search box there

> one could see this as the search for a module name

.

For me pressing Home and Enter to be able to do an exact search is very weird.  Why not use quotes: "startswith%"

.

In the tooltip there's a typo ('serach instead of search'), and I would also replace 'text' with 'photos'.

My proposal for the tooltip would be:

filter photos by metadata, tags and filename.
per default wildcards are auto-applied.
use quotes for an exact search (with `%' as custom wildcard).

phweyland commented:
> I totally agree, that 'search photos' would be better.

Would you mean that the other filters (stars, colors, ...) have a different purpose ? (not to search images ?)
I've already answered this topic above. Disagreed.

The whole purpose of the quick filter bar is to filter the images shown in lighttable. 
Each filter is specific in the way it works.
As such, I would change the label "view" to "filter" or something equivalent (idem for its tooltip).
The stars filter tooltip could also be more specific.

> Why not use quotes: "startswith%"

I don't like this better.
I would have preferred Ctrl+Enter but I've been a bit lazy to code the few extra lines ...
But I have no strong opinion here.
What are the preferences of the others ? (Home, Enter; Ctrl+Enter; "xxx") ?

> In the tooltip there's a typo ('serach instead of search')

Thanks I'll fix this.

quovadit commented:
> Would you mean that the other filters (stars, colors, ...) have a different purpose ?

no, of course all filters relate to photos, but it's obvious that a color-label does not filter module names.

I like the idea of changing 'view' to 'filter'.
So maybe we could then just remove the label 'search text'? (or just say 'search')

Nilvus commented:
> > I totally agree, that 'search photos' would be better.
> 
> Would you mean that the other filters (stars, colors, ...) have a different purpose ? (not to search images ?) I've already answered this topic above. Disagreed.
> 
> The whole purpose of the quick filter bar is to filter the images shown in lighttable. Each filter is specific in the way it works. As such, I would change the label "view" to "filter" or something equivalent (idem for its tooltip). The stars filter tooltip could also be more specific.

Of course, it's the whole purpose of the quick filter bar. And all search on images. I understand that you can find search photos (or search on photos) not useful but your arguments works so also for search text. A search box is to write text, so a search box do not have different purpose, than search text. Disagreed to and already explained too...

I agree to filter instead of view. And @quoavidt last comment is probably the simple and easiest choice. Just put "search" in search box if you don't want search photos. Search text is really strange for a search box (just evident that it's for searching text). And it's better to let some text just to make the search box more visible/evident.

phweyland commented:
The idea which should pop up is "search in all text fields"... "search text" is for the time being the best option I have in that a short field.

quovadit commented:
it's a text-field, so you can only enter text by definition, but this discussion seems wasteful.

More important would be:  Press 'Home-Button' + Enter to search without wildcards
Is this ok for everybody?

phweyland commented:
Instead of waiting the first 3 characters entered, uses 2 timeouts to start the query: 0.4 sec after each key or 1,5 sec of sequential keys. Seems to smooth a bit more the query execution versus the key entry. 

Some measurements of the collection query on my machine: with 55K images 2 sec with no filter and 2,6 sec with a text filter. Not so huge difference.

Included the @quovadit suggestion for exact search. Each '"' disables the corresponding wildcard.

Updated the different filter tooltips to turn them consistent. Also removed the disputed placeholder text.

quovadit commented:
can't test at the moment, but this sounds great, thanks a lot!

Nilvus commented:
Better, thanks. There's 2 issues with last update anyway:

- first and most annoying: it's badly aligned vertically (combobox again). See:
![image](https://user-images.githubusercontent.com/45535283/154809742-08cf3b1f-ea19-4f04-902a-b5ca314d5f48.png)

- and as you can see, "filter" text is translated, in french, as "filtre" automatically (as filter already exist on another part). This could probably be updated in po files but maybe starting with "filter:" (so by adding colon after filter) instead of "filter" would be more appropriate to help translation too and differentiate it from other "filter" text.

phweyland commented:
@Nilvus, thanks for testing.

Mine is not perfect but not as bad as yours.
![image](https://user-images.githubusercontent.com/23012047/154819807-11cdaa8e-9f7f-48da-8b07-12288751dd09.png)
![image](https://user-images.githubusercontent.com/23012047/154819828-9ac0b92a-d894-42d1-9e92-b6381bfe39b2.png)

Should be great to have this solved in a generic manner, without having each time to tweak margin and padding ! :smile: 

> maybe starting with "filter:"

Nilvus commented:
> Should be great to have this solved in a generic manner, without having each time to tweak margin and padding ! smile EDIT. I think that this combox is not vertically centered by construction. It should as the other field types.

I agree. That's again the nightmare of bauhaus widget. All bauhaus need to be removed and replace with more standard Gtk, starting by those bauhaus_combobox.
 
> > maybe starting with "filter:"
> 
> EDIT (sorry posted too early) I've tested this with "view:". Now, with the visual separation of the fields, I don't feel the need of this any more. But that is easy to fix if necessary. And also uncommon in dt.

You mean that colon is not necessary. Probably, but standard translation is not good. That will need to be fixed. If I understand you correctly here, you mean so fix it on translation.

phweyland commented:
> but standard translation is not good.

It was not wrong for me... you probably prefer the verb.
Anyway, yes, there is an adhoc mean to translate a text differently in different contexts. Remain to find it again ... 

> That's again the nightmare of bauhaus widget. All bauhaus need to be removed and replace with more standard Gtk, starting by those bauhaus_combobox.

At the end, everything is gtk, even bauhaus_combobox. It should be possible to make it vertically centered, shouldn't it ?

Nilvus commented:
> It was not wrong for me... you probably prefer the verb. Anyway, yes, there is an adhoc mean to translate a text differently in different contexts. Remain to find it again ...

Of course, "filtre" is not wrong but it remains wrong as it's not consistent with "trier par" which is a verb. So a verb is needed!

> At the end, everything is gtk, even bauhaus_combobox. It should be possible to make it vertically centered, shouldn't it ?

Of course, it's gtk but what I mean is classic gtk. To be more precise, there's default combobox with gtk (like those in preferences dialog window. And they don't need specifically made one with more complex code like bauhaus. Actually, we have 2 different ways to create combobox and they are not consistent. All combobox should be made same way, with default combobox proposed by Gtk code. That what I mean here.
Actual bauhaus combobox could probably be vertically centered with CSS. I didn't test by now but this is more a workaround for me than the best solution. Could be made waiting for better of course. 

Said that, I just wanted to remember you that was good on a previous commit of that PR. As I remember, should be commit 6 days ago (15th february) or previous one.

phweyland commented:
> that was good on a previous commit of that PR

For me it was not better before.

I've simplified a bit the code, and make the different elements collaborate better height wise (width wise, @dterrahe had already solved the issues taking in account the actual needed width).

The goal was to create boxes, with a background a bit different than the one of the filter bar, vertically aligned, with the same height. This works like a charm removing the tweaking code and css attached to these elements.
The highest element drives the height of all the boxes without any specific tweak.
Now this works in a better manner, width and height wise, I don't say this is perfect!
Remains a barely noticeable (on my screen) vertical misalignment of the text 'inside' the combobox, which is more embarrassing on your screen.

> Actual bauhaus combobox could probably be vertically centered with CSS.

Any idea I could try about this ?

We have to discover which setting does drive the current behavior and to fix it.
That may help to understand what makes the difference between our screens.
Here are the darktable and darktable-elegant-grey displays on a fresh session:
![image](https://user-images.githubusercontent.com/23012047/155011348-ec7601bc-a5cc-46a0-84b6-2ff03d1b854a.png)
![image](https://user-images.githubusercontent.com/23012047/155011452-4e34fd42-9ebe-499b-b20e-357e424d755b.png)
What are yours ?

Nilvus commented:
> > that was good on a previous commit of that PR
> 
> For me it was not better before.

Hmm, maybe I've had some CSS code at that time. I've worked with some but erase some code parts. Unfortunately...

> I've simplified a bit the code, and make the different elements collaborate better height wise (width wise, @dterrahe had already solved the issues taking in account the actual needed width).

And that's good.

> The goal was to create boxes, with a background a bit different than the one of the filter bar, vertically aligned, with the same height. This works like a charm removing the tweaking code and css attached to these elements. The highest element drives the height of all the boxes without any specific tweak. Now this works in a better manner, width and height wise, I don't say this is perfect! Remains a barely noticeable (on my screen) vertical misalignment of the text 'inside' the combobox, which is more embarrassing on your screen.

Maybe with some "standard" combobox, this would be better. But it's probably a lot more work to remove the use of bauhaus one to replace one with more standard one.
 
> > Actual bauhaus combobox could probably be vertically centered with CSS.
> 
> Any idea I could try about this ?
> We have to discover which setting does drive the current behavior and to fix it. That may help to understand what makes the difference between our screens.

I suggest to let me search that and add that in a PR I'm working on to continue improve some details and CSS code.

phweyland commented:
The word "filter" has now a translation context ("quickfilter") for adhoc translation.

Nilvus commented:
I think this is now ready to be merged? What do you think about that @phweyland? If it's ok, so ready to review for @TurboGit.

phweyland commented:
Yes, I think also this is ready for Pascal review.
Thanks for testing and suggestions.

TurboGit reviewed (approved):
Works for me. Thanks!

By any mean, if it is not too difficult to add the image n° into the text filter it would be nice.

dterrahe commented:
Apologies for not paying more attention to this PR before it was merged. I now (after rebasing my local work) notice that the "filter" and "sort by" comboboxes had their labels moved to an external widget. This has a few consequences which may not all have been intentional. 

For one, the shortcuts are set up by `dt_bauhaus_widget_set_label` and therefore no longer exist. 

For another, one of (in my opinion) advantages of bauhaus is that for narrow widgets, it prioritises the selected item/value over the label when ellipsizing. Since the new header has a lot of additional content and there were already issues with it not fitting before (addressed by #10791) this is not a hypothetical case. With the separate label, it will either never ellipsize (as now) or sometimes ellipsize even when the currently selected dropdown value is narrow enough to see both. I much prefer to see the actual value (since the label is always the same and basically described in the tooltip anyway).

The discussion above is long so not easy to find the initial rationale for separating the label, but is it still valid since #10791?

phweyland commented:
> This has a few consequences which may not all have been intentional.

They haven't.
I think there should not be any issue (I hope) to revert this part.
If the case of `sort by` is clear, the case of filter is a bit different because now we have several filters, may be more tomorrow, and they will be selectable. That will interfere as well with shortcuts you mention.
Anyway, the discussion is opened.

dterrahe commented:
> I think there should not be any issue (I hope) to revert this part.

Would you like me to have a closer look (and prepare a PR)? I won't immediately get round to it, but there is time before 4.0...

I have missed what the plans for "filter" are. Is the intention to also merge non-rating filters into the same dropdown? Giving a thought to inputng when implementing this (please @ me if/when you need help) would be nice. It doesn't just allow assigning shortcuts but also means instant access from lua (via `dt_action_process`). Although for something as central as filtering a dedicated api might eventually still be nicer.

phweyland commented:
Please do. Just note that the filters are now three, so the label is not any more the stars one.

AlicVB commented:
I would add that the plan for filters are (basically) : 
- use more "friendly" widgets
- allow users to set which filters to show/hide
- allow users to use almost all what is actually proposed in collect as filters (with different widgets and special applying...)

you can eventually see my WIP (just a draft actually) in my PR #11066 . (Note that once things will have settled down there, I will certainly ask you for help about the shortcuts :))

in this context, I would argue like @phweyland  that the "filter" needs to be separated.

dterrahe commented:
> in this context, I would argue like @phweyland that the "filter" needs to be separated.

Understood and agreed; thanks both.

But now I wonder if the "stars" dropdown should have a label "ratings". Because "all" doesn't mean "filter=all" (because there could still be a filter on color labels or text), it just means "all ratings"...
